nm: Preserve current IP setting for multiconnect profile

cathay4t opened the pull request:
Given a interface is activated by a NM profile of multiconnect
When applying empty state for this interface,
Then this interface will get IPv4 and IPv6 disabled.

This is because current nmstate code treat multiconnect or NM connection
without interface name as not matched, hence `exist_nm_conn` is set to
None which lead to nmstate discarding current IP settings.

Cloning multiconnect profiles could have risk when NM added more
properties impacting activation but nmstate not aware.

Hence we create NM connection from scratch in this case by only
preserving IP settings from multiconnect. Other non-ip settings applied
by multiconnect profile will be discard if not mentioned in desired
state.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-61890

ihuguet commented:
If I understand it well, if a multiconnect profile is active on the interface, it doesn't match. Then this just creates a new profile using the current applied ipv4 and ipv6 settings. But what about all the other properties of the profile? I would expect that they are preserved as well, if I don't define them in the desired state.

Why not this other approach: first check if a multiconnect profile is active on the interface. If there is one, as we cannot just modify this profile, because it would affect to other devices too, clone it so we create a new profile with the same settings, but multiconnect=false.

cathay4t commented:
> If I understand it well, if a multiconnect profile is active on the interface, it doesn't match. Then this just creates a new profile using the current applied ipv4 and ipv6 settings. But what about all the other properties of the profile? I would expect that they are preserved as well, if I don't define them in the desired state.
> 
> Why not this other approach: first check if a multiconnect profile is active on the interface. If there is one, as we cannot just modify this profile, because it would affect to other devices too, clone it so we create a new profile with the same settings, but multiconnect=false.

We see many race problem in nmstate 1.x python code base if we try to get active profile using `org.freedesktop.NetworkManager.Device.ActiveConnection` interface, so there is not reliable way to know a activate profile of a interface.
If we try to match a existing profile by mimicing NM matching methods, it would be way too complex.

Nmstate has no intention to support user who use multiconnect profiles for non-ip settings. I heard no nmstate user complain about this.

cathay4t commented:
Meanwhile, NM also match NM connection even mutliconnect disabled as long as that connection has no matching method, it will be used for first interface. Cloning that profile instead of creating from scratch would cause too many problems for example `match` leftovers.

Unless a valid use case request us to properly support non-ip settings from multiconnect, I would like to limit our scope on this.

ffmancera reviewed (approved):
As we want to limit the scope of the MR, it looks good to me.

ihuguet commented:
If it's decided, please clarify in the comment that this affects to "multiconnect" profiles and that, as a consequence, other things configured in the profile are ignored.

ihuguet commented:
Anyway, I don't fully understand the use case. If nmstate doesn't support multiconnect profiles, why are there multiconnect profiles active and being used? You mentioned that it's an issue for OCP. Why is OCP using multiconnect profiles? Why then they try to use nmstate later?

cathay4t commented:
OCP(OpenShift) is installing a multiconnect connection files to `/run/NetworkManager/system-connections` folder by enabling ipv4.method auto and ipv6.method auto on ALL interfaces during installation stage(ram initfs). It is new environment nmstate try to fit it. The use case is nmstate should preserve current IP settings even IP is not mentioned in desired state.

Nmstate has no property indicate the support of this multiconnect, but can works when mutliconnect been created by other tools.

ihuguet commented:
OK, got it, seems a valid use case.

I still think that it's easy and better the other approach:
- If there is an active connection and it's multiconnect: create a new NMConnection copying the active one, but with multiconnect=no (this PR's approach is to create an empty/default NMConnection and copy the ipv4 and ipv6 parts only).
- Otherwise, proceed as we are doing now.

But as I said, I won't insist if both of you are OK with this approach. Just add the comments that I said, please.

cathay4t commented:
Added these lines to commit message:

```
Cloning multiconnect profiles could have risk when NM added more
properties impacting activation but nmstate not aware.

Hence we create NM connection from scratch in this case by only
preserving IP settings from multiconnect. Other non-ip settings applied
by multiconnect profile will be discard if not mentioned in desired
state.
```

cathay4t committed:
nm: Preserve current IP setting for multiconnect profile

Given a interface is activated by a NM profile of multiconnect,
When applying empty state for this interface,
Then this interface will get IPv4 and IPv6 disabled.

This is because current nmstate code treat multiconnect or NM connection
without interface name as not matched, hence `exist_nm_conn` is set to
None which lead to nmstate discarding current IP settings.

Cloning multiconnect profiles could have risk when NM added more
properties impacting activation but nmstate not aware.

Hence we create NM connection from scratch in this case by only
preserving IP settings from multiconnect. Other non-ip settings applied
by multiconnect profile will be discard if not mentioned in desired
state.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-61890

Signed-off-by: Gris Ge <fge@redhat.com>
